### PR TITLE
IGNITE-17533 Implement a meta B+Tree of indexes

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-17533

### CODE REVIEW

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMeta.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.ignite.internal.tostring.S;
+
+/**
+ * Index tree meta information.
+ */
+public class IndexMeta {
+    @IgniteToStringInclude
+    private final UUID id;
+
+    private final long rootPageId;
```
Comment (SammyVimes) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMeta.java:31]: Why not add root page id for logging purposes?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+     *
+     * @param ver Page format version.
+     */
+    private IndexMetaInnerIo(int ver) {
+        super(T_INDEX_META_INNER_IO, ver, true, SIZE_IN_BYTES);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void store(long dstPageAddr, int dstIdx, BplusIo<IndexMeta> srcIo, long srcPageAddr, int srcIdx) {
```
Comment (SammyVimes) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java:48]: why override if we call a default method from the interface?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        IndexMetaIo.super.store(dstPageAddr, dstIdx, srcPageAddr, srcIdx);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void storeByOffset(long pageAddr, int off, IndexMeta row) {
+        IndexMetaIo.super.storeByOffset(pageAddr, off, row);
```
Comment (SammyVimes) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java:55]: Same question about override

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+     * @param ver Page format version.
+     */
+    private IndexMetaLeafIo(int ver) {
+        super(T_INDEX_META_LEAF_IO, ver, SIZE_IN_BYTES);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void store(long dstPageAddr, int dstIdx, BplusIo<IndexMeta> srcIo, long srcPageAddr, int srcIdx) {
+        IndexMetaIo.super.store(dstPageAddr, dstIdx, srcPageAddr, srcIdx);
```
Comment (SammyVimes) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java:49]: Same question about override

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void storeByOffset(long pageAddr, int off, IndexMeta row) {
+        IndexMetaIo.super.storeByOffset(pageAddr, off, row);
```
Comment (SammyVimes) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java:55]: Same question about override

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaIo.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    int INDEX_ID_MSB_OFFSET = 0;
+
+    /** Offset of the {@link UUID#getLeastSignificantBits() least significant bits} of the index ID (8 bytes). */
+    int INDEX_ID_LSB_OFFSET = INDEX_ID_MSB_OFFSET + Long.BYTES;
+
+    /** Index root page ID offset - long (8 bytes). */
+    int INDEX_ROOT_PAGE_ID_OFFSET = INDEX_ID_LSB_OFFSET + Long.BYTES;
+
+    /** Payload size in bytes. */
+    int SIZE_IN_BYTES = INDEX_ROOT_PAGE_ID_OFFSET + Long.BYTES;
```
Comment (SammyVimes) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaIo.java:49]: It's a matter of taste, but I don't really like that size is calculated from offset. But I'll leave it up to you

**Review by SammyVimes — CHANGES_REQUESTED** [review]

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMeta.java (hunk 1 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/IndexMeta.java:31]: Primitives are automatically added to logging in **S#toString**.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java (hunk 2 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java:48]: I tried to remove it, the compiler swears.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaInnerIo.java:55]: I tried to remove it, the compiler swears.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java (hunk 4 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java:49]: I tried to remove it, the compiler swears.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java (hunk 5 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaLeafIo.java:55]: I tried to remove it, the compiler swears.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaIo.java (hunk 6 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/meta/io/IndexMetaIo.java:49]: you're right, it's better to write explicitly.

**Review by SammyVimes — APPROVED** [review]
LGTM

**Pull request merged**